src/NexusMods.DataModel.SchemaVersions/SchemaFingerprint.cs

```diff
@@ -0,0 +1,32 @@
+using System.IO.Hashing;
+using System.Text;
+using NexusMods.Hashing.xxHash3;
+using NexusMods.MnemonicDB.Abstractions;
+
+namespace NexusMods.DataModel.Migrations;
+
+/// <summary>
+/// Tools for generating a hash of all the attributes of a schema so that we can detect changes.
+/// </summary>
+public class SchemaFingerprint
+{
+    public static Hash GenerateFingerprint(IDb db)
+    {
+        StringBuilder sb = new();
+        var cache = db.AttributeCache;
+        
+        foreach (var id in cache.AllAttributeIds.OrderBy(id => id.Id))
+        {
+            var aid = cache.GetAttributeId(id);
+            sb.AppendLine(id.ToString());
+            sb.AppendLine(cache.GetValueTag(aid).ToString());
+            sb.AppendLine(cache.IsIndexed(aid).ToString());
+            sb.AppendLine(cache.IsReference(aid).ToString());
+            sb.AppendLine(cache.IsCardinalityMany(aid).ToString());
+            sb.AppendLine(cache.IsNoHistory(aid).ToString());
+            sb.AppendLine("--");
```

Comment:
Should we also include whether an attribute is optional or not here?

Reply:
Optional doesn't make a difference to the database itself, so it's not important here, we can add it later if we need to.

Reply:
Changing an attribute from optional to required is something we want to know about. That's the only operation (add, remove, modify) on optional attributes that can result in issues between schema versions.

While the DB probably doesn't care, it's the usage I'm worried about. Going from optional to required means our code now assumes that entities will always have the attribute. If we don't catch this change in our verify test and don't update our code or write a migration to account for this change, we'll end up getting runtime exceptions.

If the schema changes, an attribute from optional to required, we need to write a migration to fill in the blanks with some form of data.

Reply:
See below